Store EC2 instance pricing in DB

thibautcornolti opened the pull request:
Fetch EC2 Pricing thanks to the AWS API and store it in Database.

thibautcornolti committed:
Store ec2 instance pricing in db

TRAC-594 #comment Store ec2 instance pricing in db

thibautcornolti committed:
Add comments to awsec2product.go

TRAC-594 #comment Add comments to awsec2product.go

thibautcornolti committed:
Fix logger in awsec2product.go

TRAC-594 #comment Fix logger in awsec2product.go

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+var (
+	// storeAttribute contains the reference between
+	// a field name and the way to store its value.
+	storeAttribute = map[string]func(string, interface{}, *models.AwsProductEc2, jsonlog.Logger){
```
`func(string, interface{}, *models.AwsProductEc2, jsonlog.logger)` is not a descriptive type name, why not alias it with something more descriptive, such as `attributeStoring` ?

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+const (
+	// urlEC2Pricing is the URL used by downloadJSON to
+	// fetch the EC2 pricing.
+	urlEC2Pricing = "https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/index.json"
```
You should put this url inside of a config file instead of putting it directly into the code

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+// ID in the struct which will be imported in the database.
+func storeAttributeRegion(name string, value interface{}, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger) {
+	dbAwsRegion, err := models.AwsRegionByPretty(db.Db, value.(string))
+	if err != nil {
```
You check if `err != nil` but does it only mean that the region is not created inside the database ?
You do not check if it could be another type of error

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+		newDbAwsRegion.Pretty = value.(string)
+		newDbAwsRegion.Region = value.(string)
+		if err := newDbAwsRegion.Insert(db.Db); err != nil {
+			logger.Error("Error during dbAwsRegion inserting", err)
```
You log an error but do not return it

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+
+// storeAttributeGenericBool stores an bool value in the struct
+// which will be imported in the database.
+func storeAttributeGenericBool(name string, value interface{}, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger) {
```
Your function does not return anything.
Please try to avoid this model as it makes for unmaintainable code quickly.

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+
+// storeAttributeGenericString stores an string value in the struct
+// which will be imported in the database.
+func storeAttributeGenericString(name string, value interface{}, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger) {
```
Ditto here for a function that does not return anything

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+	defer reader.Close()
+	body, err := ioutil.ReadAll(reader)
+	if err != nil {
+		logger.Error("Body reading failed", err)
```
You log an error but do not return it or even return from the function

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+	var bodyMap map[string]interface{}
+	if err = json.Unmarshal(body, &bodyMap); err != nil {
+		logger.Error("JSON Unmarshal failed", err)
+		return
```
You log an error but do not return it

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+				fct(name, attribute, &dbAwsProduct, logger)
+			}
+		}
+		dbAwsProduct.Insert(db.Db)
```
As stated in the function prototype : `func (ape *AwsProductEc2) Insert(db XODB) error {` this method can return an error but you do not check for it
This also means that this functions must return an error

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+
+	// run query
+	models.XOLog(sqlstr)
+	db.Db.Exec(sqlstr)
```
I haven't checked the `mysql` go package, but I am pretty sure that a `Exec` statement can generate an error, but you do not check for it

schubev replied:
Please do your SQL queries only in the `models` package, which abstracts all database access. Since we don’t really have an ORM this is how we abstract storage.

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+	logger := jsonlog.DefaultLogger
+	res, err := downloadJSON(logger)
+	if err != nil {
+		logger.Error("Error during JSON downloading", err)
```
You log an error but do not return it

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+		purgeCurrent()
+		importResult(res, logger)
+	}
+	return nil, nil
```
Why do you need the `interface{}` return type, if you will always return `nil` ?

schubev replied:
In that case, why is there even a return value ?

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+		importResult(res, logger)
+	}
+	return nil, nil
+}
```
Please write Unit Tests

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+
+	"github.com/trackit/jsonlog"
+	"github.com/trackit/trackit2/db"
+	"github.com/trackit/trackit2/models"
```
Please respect organize your imports in three sections : standard library imports, imports from other projects (not `github.com/trackit/trackit2`) and imports from the current project.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+// storeAttributeRegion creates the region in the database
+// if it doesn't exist. Then, the function stores the region's
+// ID in the struct which will be imported in the database.
+func storeAttributeRegion(name string, value interface{}, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger) {
```
What does `value` represent ? Neither the name nor the type mean anything.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+func storeAttributeGenericBool(name string, value interface{}, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger) {
+	if value, ok := value.(string); ok {
+		var val int
+		if "Yes" == value {
```
Yoda conditions not used we have. By the current style abide you shall.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+		if "Yes" == value {
+			val = 1
+		}
+		switch name {
```
Why do you add a single-cased `switch` when you just used an `if` for the same purpose ?

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+// importResult parses the body returned by downloadJSON and
+// inserts the pricing in the database.
+func importResult(reader io.ReadCloser, logger jsonlog.Logger) {
+	defer reader.Close()
```
Good !

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,191 @@
+	}
+	lastFetch, err := models.AwsFetchPricingByProduct(db.Db, "ec2")
+	if err != nil {
+		req.Header.Add("If-None-Match", lastFetch.Etag)
```
It’s good that you make sure we don’t process something we already have processed, though perhaps you should extract this in its own function. This one is very long.

schubev reviewed (dismissed):
# JSON parsing
Go’s `encoding/json` is very good at parsing data when you know what it looks like, which you do here. By defining some types you can get `encoding/json` to handle the parsing in a typesafe manner. For example:
```go
type PricingSku string

type Ec2Princing struct {
    PublicationDate time.Time
    Products map[PricingSku]PricingProduct
}

type PricingProduct string {
  Sku PricingSku
  ProductFamily string
  Attributes Ec2ProductAttributes
}

type Ec2ProductAttributes struct {
  InstanceType string
  InstanceFamily string
  Memory string
  Storage string
}
```

# Memory usage
As of this review, the document is 172MiB and your current method stores it all in memory. Please try to use streaming when working with such data. https://golang.org/pkg/encoding/json/#example_Decoder_Decode_stream shows how you can decode a JSON document as a stream without storing it all.

# Lack of transactions
You do not use an SQL transaction for this job. This means that in case of error you will leave the database in a dirty state. Please use a transaction to commit your changes to storage only once the whole job is finished.

# Lack of context
Please use `context.Context`, this allows several things : a logger can be stored there which will add some ID to all log events for correlation ; it can store the SQL transaction ; it lets the job be cancelled cleanly.

thibautcornolti committed:
Designed config.go file, refactored the EC2 pricing fetching from AWS.

Using of struct to decode the downloaded json instead of map
Decoding from io.ReaderCloser instead of String
Using context and transactions

TRAC-594

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+	}
+)
+
+func storeRegionID(region string, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger, sqlTx models.XODB) error {
```
`sqlTx` is just named `tx` everywhere else in the project. Why change the naming convention ?

schubev replied:
You modify the arguments while neither documenting the function nor reflecting that fact in the function name. Don’t.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+func storeRegionID(region string, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger, sqlTx models.XODB) error {
+	dbAwsRegion, err := models.AwsRegionByPretty(sqlTx, region)
+	if err != nil && err.Error() == "sql: no rows in result set" {
+		var newDbAwsRegion models.AwsRegion
```
This is clearer as
```go
var newDbAwsRegion = models.AwsRegion{
    Pretty = region,
    Region = region,
}
```

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+		newDbAwsRegion.Pretty = region
+		newDbAwsRegion.Region = region
+		if err := newDbAwsRegion.Insert(sqlTx); err != nil {
+			logger.Error("Error during dbAwsRegion inserting", err.Error())
```
If I were to see this in the logs, I’d have no idea what it means. When logging, ensure your message is meaningful outside the code.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+	} else if err == nil {
+		dbAwsProduct.RegionID = dbAwsRegion.ID
+	} else {
+		logger.Error("Error during dbAwsRegion fetching", err.Error())
```
Same here, I’d have no idea what that means if I saw it pop in my logs.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+
+func storeRegionID(region string, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger, sqlTx models.XODB) error {
+	dbAwsRegion, err := models.AwsRegionByPretty(sqlTx, region)
+	if err != nil && err.Error() == "sql: no rows in result set" {
```
Please directly compare to `sql.ErrNoRows`.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+)
+
+type (
+	PricingSku string
```
An SKU is an SKU, you don’t need to say it’s a “Pricing SKU”.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+// to models.AwsProductEc2
+func storeAttributes(attributes *PricingAttribute, dbAwsProduct *models.AwsProductEc2, logger jsonlog.Logger, sqlTx models.XODB) (err error) {
+	dbAwsProduct.InstanceType = attributes.InstanceType
+	if attributes.CurrentGeneration == "Yes" {
```
Do you expect only the two strings “Yes” and “No” ? Then make sure an error is logged if we get an unexpected string.

schubev commented on docker/sql/schema.sql:
```diff
@@ -52,3 +52,38 @@ CREATE TABLE aws_bill_repository (
 	SELECT * FROM aws_bill_repository WHERE next_update <= NOW()
 ;
+
+CREATE TABLE aws_fetch_pricing (
```
What is an AWS fetch pricing ?

schubev commented on docker/sql/schema.sql:
```diff
@@ -52,3 +52,38 @@ CREATE TABLE aws_bill_repository (
+	sku VARCHAR(255) NOT NULL,
+	region_id INTEGER NOT NULL,
+	instance_type VARCHAR(255) NOT NULL,
+	current_generation INTEGER(1) NOT NULL,
```
Use a boolean.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+	dbAwsRegion, err := models.AwsRegionByPretty(sqlTx, region)
+	if err != nil && err.Error() == "sql: no rows in result set" {
+		var newDbAwsRegion models.AwsRegion
+		newDbAwsRegion.Pretty = region
```
Why the extra “pretty” field if it’s just the same value ? This seems useless and the sole justification for creating a new table for regions. Adding a join is adding complexity, and it seems we do not need it at all.

schubev commented on config/config.go:
```diff
@@ -44,18 +44,87 @@ var (
-	flag.StringVar(&BackendId, "backend-id", "", "The ID to be sent to clients through the 'X-Backend-ID' field. Generated if left empty.")
-	flag.StringVar(&EsAuthentication, "es-auth", "basic:elastic:changeme", "The authentication to use to connect to the ElasticSearch database.")
-	flag.StringVar(&EsAddress, "es-address", "http://127.0.0.1:9200", "The address of the ElasticSearch database.")
-	flag.BoolVar(&PrettyJsonResponses, "pretty-json-responses", false, "JSON HTTP responses should be pretty.")
```
No.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+	if val, err := strconv.Atoi(attributes.Ecu); err == nil {
+		dbAwsProduct.Ecu = val
+	}
+	if attributes.LocationType == "AWS Region" {
```
Again, please ensure unexpected values, for example for new types of locations, throw errors so that we may be notified the code needs updating.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+	if attributes.CurrentGeneration == "Yes" {
+		dbAwsProduct.CurrentGeneration = 1
+	}
+	if val, err := strconv.Atoi(attributes.Vcpu); err == nil {
```
You should at the very least log this kind of error.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+		return nil, err
+	}
+	if res.StatusCode == 304 {
+		logger.Info("JSON Downloading halted: Already exists with ETag.\n", lastFetch.Etag)
```
`"JSON download halted:…"`

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+	}
+	lastFetch, err := models.AwsFetchPricingByProduct(db.Db, "ec2")
+	if err == nil {
+		req.Header.Add("If-None-Match", lastFetch.Etag)
```
This is very good !

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+	if err != nil {
+		return nil, err
+	}
+	lastFetch, err := models.AwsFetchPricingByProduct(db.Db, "ec2")
```
Please extract `"ec2"` into a constant. This is generally good practice, because a typo in a string can be hard to debug, while a typo in an identifier usually immediately causes a compile error.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+// checking if there's already the same version in the database.
+func downloadJSON(logger jsonlog.Logger, sqlTx models.XODB) (io.ReadCloser, error) {
+	hc := http.Client{}
+	req, err := http.NewRequest("GET", config.UrlEC2Pricing, nil)
```
Please use `http.MethodGet` instead of a literal string.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode == 304 {
```
Please use `http.StatusNotModified`, not everyone know HTTP codes at the top of their head.

schubev replied:
Also, see my comment about the `"ec2"` string, it applies here as well.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+		return nil, nil
+	}
+	if err = saveLastFetch(lastFetch, res.Header["Etag"][0], sqlTx); err != nil {
+		logger.Error("Error when saving the dbAfp", err.Error())
```
Again, ensure your logs are meaningful outside of the code that produce them.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+// ImportEC2Pricing downloads the EC2 Pricing from AWS and
+// store it in the database.
+func ImportEC2Pricing() error {
+	logger := jsonlog.DefaultLogger
```
Please use a context and get your logger from there.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+// store it in the database.
+func ImportEC2Pricing() error {
+	logger := jsonlog.DefaultLogger
+	sqlTx, err := db.Db.BeginTx(context.Background(), nil)
```
The context given to this function should also be used here instead of the background context.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+	logger := jsonlog.DefaultLogger
+	sqlTx, err := db.Db.BeginTx(context.Background(), nil)
+	if err != nil {
+		logger.Error("Error when beginning sqlTx", err.Error())
```
Ensure your logs are meaningful without the code that produced them. I won’t make this comment again, check all your logs.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+	if err == nil {
+		req.Header.Add("If-None-Match", lastFetch.Etag)
+	}
+	res, err := hc.Do(req)
```
Use a context.

schubev commented on config/config.go:
```diff
@@ -44,18 +44,87 @@ var (
+		&UrlEC2Pricing,
+		"url-ec2-pricing",
+		"https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/index.json",
+		"The URL used to download the EC2 pricing",
```
You forgot the dot.

schubev commented on docker/sql/schema.sql:
```diff
@@ -52,3 +52,38 @@ CREATE TABLE aws_bill_repository (
+	CONSTRAINT UNIQUE KEY (product)
+);
+
+CREATE TABLE aws_region (
```
Because you just store the same data in `region` and `pretty`, this table adds nothing and increases complexity. I think you should have a look at “Codd’s normal forms”, BTW.

schubev commented on docker/sql/schema.sql:
```diff
@@ -52,3 +52,38 @@ CREATE TABLE aws_bill_repository (
+	operating_system VARCHAR(255) NOT NULL,
+	ecu INTEGER NOT NULL,
+	CONSTRAINT PRIMARY KEY (id),
+	CONSTRAINT UNIQUE KEY (sku),
```
Why not use the SKU as a primary key ? You can use strings as primary keys. Also, if you can find documentation that SKU’s have a fixed, number of characters, you can even use `CHAR` instead of `VARCHAR`.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+
+	decoder := json.NewDecoder(reader)
+	var bodyMap EC2Pricing
+	if err := decoder.Decode(&bodyMap); err != nil {
```
Please stream the JSON document as per my previous review. This document is known to be quite big and reasonable memory footprint is a requirement for this project.

schubev commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,179 @@
+		logger.Error("Error when beginning sqlTx", err.Error())
+		return err
+	}
+	res, err := downloadJSON(logger, sqlTx)
```
Here, if there are no errors, you
1. Open a connection to download a JSON document
2. Purge a table in the database
3. Consume the JSON document

You can’t know for sure that the database purge will end before the connection times out, so you have two possible solutions :
1. Insert products along with the etag (or some other id) for the document they came from, with an `INSERT … ON DUPLICATE KEY UPDATE`, then remove from the database all records with another etag (this is my preferred option).
2. Immediately consume the JSON document by writing it to a temporary file, purge the database, then consume the temporary file (don’t forget to remove it afterwards).

schubev reviewed (dismissed):
You handled downloading the JSON document well by checking the ETag and errors.

However, in my previous I asked you to consume the JSON document as a stream, which you didn’t.

Also, please use contexts wherever you perform long-running, blocking actions such as database requests (you can attach the context to the transaction) or HTTP requests. Contexts allow jobs to be cancelled, and carry values useful for logging.

Finally, with SQL you can insert more than one record at a time. You perform an SQL request for each and every SKU, which means an additional network roundtrip everytime. This can lead to poor performances. Please batch your inserts when they are numerous.

thibautcornolti committed:
Stream JSON instead of download & Bulk insert products

TRAC-594

giubil commented on aws/product/awsec2product.go:
```diff
@@ -0,0 +1,226 @@
+// to models.AwsProductEc2
+func storeAttributes(ctx context.Context, attributes *Attribute,
+	dbAwsProductPricing *models.AwsProductPricingEc2, tx models.XODB) {
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
```
You take `tx models.XODB` as a parameter, but you do not use it.

schubev reviewed (approved):
The SQL insert part is clunky, but with the tools we currently use I don’t really know how we can do better.
